### PR TITLE
App timelines

### PR DESCRIPTION
**Pull request by Kamkalet**

### CODE REVIEW

File: src/main/java/monitoringsystemturbo/Presentation.java [hunk 1]
```diff
@@ -120,7 +120,7 @@ private static void exportData(List<String> programNames, TrackingService tracki
     private static void saveProcessesToConfig(String[] processNames) throws FileNotFoundException {
         List<Application> listToSave = new ArrayList<Application>();
         for (String processName : processNames) {
-            listToSave.add(new Application(processName, "C://"));
+            listToSave.add(new Application(processName, "C://Users/kkal6/AppData/Local/Google/Chrome SxS/Application/chrome.exe"));
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/Presentation.java:123]: maybe there is a way to customize it? Hardcoded path, especially with user id in code, is not a good idea. Consider using gradle if possible

File: src/main/java/monitoringsystemturbo/history/TimelineJsonDeserializer.java [hunk 2]
```diff
@@ -42,7 +42,7 @@ private void deserializePeriods() throws ClassNotFoundException {
                 default:
                     throw new ClassNotFoundException("Period subclass not found (" + periodClassName + ")");
             }
-            periods.add(period);
+            observableList.add(period);
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/history/TimelineJsonDeserializer.java]: change name to be more meaningful, we know that it is a observableList, but nothing about why it exists and what it stores

File: src/main/java/monitoringsystemturbo/model/app/ApplicationMonitor.java [hunk 3]
```diff
@@ -45,6 +45,7 @@ public void run() {
                 }
             }
         }
+
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/model/app/ApplicationMonitor.java]: remove eneter pls

File: src/main/java/monitoringsystemturbo/model/timeline/Timeline.java [hunk 4]
```diff
@@ -1,30 +1,32 @@
...
-import java.util.List;
+import java.util.*;
 
 public class Timeline {
 
     protected Date datetimeStart;
     protected Date datetimeEnd;
     private ApplicationState lastApplicationState = ApplicationState.NOT_RUNNING;
-    protected ArrayList<Period> periods = new ArrayList<>();
+    protected ObservableList<Period> observableList;
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/model/timeline/Timeline.java]: same thing with naming. Readability is better when name indicates what it's stored in a list

File: src/main/java/monitoringsystemturbo/model/timeline/Timeline.java [hunk 5]
```diff
@@ -37,11 +39,23 @@ public Timeline(List<ComputerStatistics> computerStatistics) {
...
 
+    public Timeline(ComputerStatistics computerStatistic) {
+        observableList = FXCollections.observableArrayList();
+
+        this.datetimeStart = computerStatistic.getSystemStartTime();
+        this.datetimeEnd = computerStatistic.getSystemCloseTime();
+
+        Period period = new RunningPeriod(datetimeStart, computerStatistic.getSystemCloseTimeProperty());
+        observableList.add(period);
+
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/model/timeline/Timeline.java]: remove enter pls

File: src/main/java/monitoringsystemturbo/presenter/MainPresenter.java [hunk 6]
```diff
@@ -60,8 +58,26 @@ public void initialize(TrackingService trackingService, MainExporter mainExporte
         applicationList.setItems(FXCollections.observableList(loadedApplications));
         renderTimelineLegend();
         initializeTimelines();
+        addCurrentTimeline();
+        addCurrentTimelineForComputer();
         initializeDatePicker();
+    }
 
+    private void addCurrentTimelineForComputer() {
+        TimelineElement timelineElement = this.timelineElements.get("Computer");
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/presenter/MainPresenter.java]: consider moving "Computer" to static constant

File: src/main/java/monitoringsystemturbo/presenter/MainPresenter.java [hunk 7]
```diff
@@ -85,7 +101,7 @@ private void initializeTimelines() {
             TimelineElement timelineElement = new TimelineElement("Computer", Arrays.asList(timeline));
             timelineElement.setTimelineViewWidthByRegion(computerTimelineContainer);
             computerTimelineContainer.getChildren().add(timelineElement);
-            timelineElements.add(timelineElement);
+            timelineElements.put("Computer", timelineElement);
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/presenter/MainPresenter.java]: it is repeated so it seems like a good idea

File: src/main/java/monitoringsystemturbo/presenter/timeline/TimelineElement.java [hunk 8]
```diff
@@ -30,10 +30,17 @@ public void setTimelineViewWidthByRegion(Region region) {
...
+        return timelineView;
+    }
+
     public void showDay(Integer day) {
         timelineView.showDay(day);
     }
 
+    public void addTimeLineModel(Timeline timeLine) {
+        timelineView.addTimeLine(timeLine);
+    }
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/presenter/timeline/TimelineElement.java:43]: add enter after this line as you added it everywhere between methods

File: src/main/java/monitoringsystemturbo/presenter/timeline/TimelineView.java [hunk 9]
```diff
@@ -107,4 +118,49 @@ public void showDay(Integer day) {
         currentDay = day;
     }
 
+    void addTimeLine(Timeline timeLine) {
+
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/presenter/timeline/TimelineView.java]: decide if you add new lines after method name or not. Almost everywhere there is no newline so I suggest not adding it, but the most import thing is to be consistent

File: src/main/java/monitoringsystemturbo/presenter/timeline/TimelineView.java [hunk 10]
```diff
@@ -107,4 +118,49 @@ public void showDay(Integer day) {
...
+                Period period = timeLine.getPeriods().get(timeLine.getPeriods().size() - 1);
+                renderPeriod(period);
+                setListenerForPeriod(period);
+            } catch (ClassNotFoundException e) {
+                e.printStackTrace();
+            }
+        }
+
+        timeLine.getPeriods().addListener((ListChangeListener<Period>) this::onNewPeriodAdded);
+
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/presenter/timeline/TimelineView.java]: same here with newline before end of a method

File: src/main/java/monitoringsystemturbo/presenter/timeline/TimelineView.java [hunk 11]
```diff
@@ -107,4 +118,49 @@ public void showDay(Integer day) {
...
+        period.getgetDatetimeEndProperty().addListener((observable, oldValue, newValue) -> {
+            double widthRatio = (double) (newValue.getTime() - period.getDatetimeStart().getTime()) / dayInMs;
+            double offsetRatio = (double) (period.getDatetimeStart().getTime() % dayInMs) / dayInMs;
+            double timelineWidth = widthProperty().getValue();
+            currentPeriodView.setWidth(timelineWidth * widthRatio);
+            currentPeriodView.setTranslateX(timelineWidth * offsetRatio);
+        });
+    }
+
+    private void onNewPeriodAdded(ListChangeListener.Change<? extends Period> c) {
```
Comment (mhabzda) [inline src/main/java/monitoringsystemturbo/presenter/timeline/TimelineView.java]: I don't like such a short name as 'c'. Please think about better, longer one

**Comment by mhabzda** [comment]
And pls resolve conflicts

File: src/main/java/monitoringsystemturbo/Presentation.java (hunk 1 above)
Comment (Kamkalet) [reply src/main/java/monitoringsystemturbo/Presentation.java:123]: It's just presentation for client thus I could skip it, or maybe should I remove entire file?

Comment (mhabzda) [reply src/main/java/monitoringsystemturbo/Presentation.java:123]: For now, let's leave it like this, we will remove it at the end

**Review by mhabzda — APPROVED** [review]
LGTM

**Pull request merged**